src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrUpdater.cs

```diff
@@ -677,6 +684,73 @@ private string GetCurrentVersion(SourceMapping mapping)
         return version.Sha;
     }
 
+    private void CleanUpRemovedRepos(string? readmeTemplatePath, string? tpnTemplatePath)
+    {
+        var deletedRepos = _sourceManifest
+            .Repositories
+            .Where(r => _dependencyTracker.Mappings.FirstOrDefault(m => m.Name == r.Path) == null)
+            .ToList();
+
+        if (!deletedRepos.Any())
+        {
+            return;
+        }
+
+        foreach (var repo in deletedRepos)
+        {
+            _logger.LogWarning("The mapping for {name} was deleted. Removing the repository from the VMR.", repo.Path);
+            DeleteRepository(repo.Path);
+        }
+
+        _fileSystem.WriteToFile(_vmrInfo.GetSourceManifestPath(), _sourceManifest.ToJson());
+
+        var versions = new AllVersionsPropsFile(_sourceManifest.Repositories);
+        var allVersionsFilePath = _vmrInfo.VmrPath / VmrInfo.GitInfoSourcesDir / AllVersionsPropsFile.FileName;
+        versions.SerializeToXml(allVersionsFilePath);
```

Comment:
Don't we have code somewhere already doing this?

Reply:
not really, something similar is in `VmrDependencyTracker` but not quite the same

Reply:
I think it should be the responsibility of the `VmrDependencyTracker`. That class is the interface that works with reading and storing data. So I think it's wrong we suddenly do it here as well.

I think we should have a method there for removal too.

Originally, I was thinking that maybe this method:
https://github.com/dotnet/arcade-services/blob/dde98af41bea7a7621175d08d6403546cf673f5d/src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrDependencyTracker.cs#L78
could work so that when you call it and update a mapping to `null` revision, it would do the remove. But I think that's too dangerous.

Reply:
When thinking about this, I think it's useful to consider situation like "what if we decided to store the versions in a completely different way". Would this behaviour be encapsulated in one class or scattered around and would require many changes across the classes?

Reply:
ok I see